any.hpp

```diff
@@ -221,20 +228,32 @@ struct static_any
 		using NonConstT = std::remove_cv_t<std::remove_reference_t<_T>>;
 		NonConstT* non_const_t = const_cast<NonConstT*>(&t);
 
-		std::array<char, _N> buff;
+		storage_t old_buff;
+		function_ptr_t old_function;
+		constexpr bool cannot_throw =
+			std::conditional<std::is_rvalue_reference<_T>::value,
+				typename std::is_nothrow_move_constructible<_T>,
+				typename std::is_nothrow_copy_constructible<_T>>::type::value;
+
+		if(!cannot_throw) {
+			old_buff = buff_;
+			old_function = function_;
+		}
+
+		destroy();
```

Comment:
Hold off on merging this just yet, I need to rethink this implementation.

One issue I just thought of is that throwing move constructors (written by insane users) will cause interesting problems in `assign`.  Everything is fine until the call to `call_copy_or_move`; what we'd really like is to call_move_if_noexcept.

More generally, making a temporary copy of the storage in 239 is invalid, because we destroy it in line 243.

Lemme think about this.

Reply:
I think the rule of thumb is that you have to abandon strong exception guarantees if the contained type does not have a nothrow copy or move constructor.